# What this file is
This is a file exploring issues that came up during building and testing. The issues are as follows

* the very first database file that was built had only 373 columns despite the an original list of columns being 470. What happened to those other columns? Which columns were dropped?

* there were to many none values within the dataset whenever the program came across a filing type that was different. Which filing type? I don't know. But I imagin that the path keys for these filings (or mulitple filing types) are different. Might it be impossible to map one filing type to another filing type?


# Edit: everything is fixed! This file is no longer needed and is only historical documentation.

# next cell should have all the side code needed

In [1]:
import os, requests, re, time, numpy as np, pandas as pd, pickle, xmltodict
from IPython.display import clear_output

def get_node_names(parent):
    node_names = []
    
    for item in parent.items():
        if item[1] != None:
            if type(item[1]) == str:
                node_names.append(item[0])
            
            elif type(item[1]) == list:
                try:
                    child_list_node_names = get_node_names(item[1][0])
                    for child in child_list_node_names:
                        string_name = item[0] + '.' + child
                        node_names.append(string_name)
                except AttributeError:
                    pass
            
            else:
                child_node_names = get_node_names(item[1])
                for child in child_node_names:
                    string_name = item[0] + '.' + child
                    node_names.append(string_name)
            
    return node_names

def get_node_values(parent):
    value_list = []
    for item in parent.items():
        if item[1] != None:
            if type(item[1]) == str:
                value_list.append(item[1])
            
            elif type(item[1]) == list:
                try:
                    child_values = get_node_values(item[1][0])
                    for value in child_values:
                        value_list.append(value)
                except AttributeError:
                    print('encountered an error')
            
            else:
                child_values = get_node_values(item[1])
                for value in child_values:
                    value_list.append(value)
    
    return value_list

In [2]:
data = pd.read_csv('affiliate_data.csv')

In [3]:
data

Unnamed: 0             Return.@xmlns  \
0              0  http://www.irs.gov/efile   
1              1  http://www.irs.gov/efile   
2              2  http://www.irs.gov/efile   
3              3  http://www.irs.gov/efile   
4              4  http://www.irs.gov/efile   
5              5  http://www.irs.gov/efile   
6              6  http://www.irs.gov/efile   
7              7  http://www.irs.gov/efile   
8              8  http://www.irs.gov/efile   
9              9  http://www.irs.gov/efile   
10            10  http://www.irs.gov/efile   
11            11  http://www.irs.gov/efile   
12            12  http://www.irs.gov/efile   
13            13  http://www.irs.gov/efile   
14            14  http://www.irs.gov/efile   
15            15  http://www.irs.gov/efile   
16            16  http://www.irs.gov/efile   
17            17  http://www.irs.gov/efile   
18            18  http://www.irs.gov/efile   
19            19  http://www.irs.gov/efile   
20            20  http://www.irs.gov/efile   
21            21  http://www.irs.gov/efile   
22            22  http://www.irs.gov/efile   
23            23  http://www.irs.gov/efile   
24            24  http://www.irs.gov/efile   
25            25  http://www.irs.gov/efile   
26            26  http://www.irs.gov/efile   
27            27  http://www.irs.gov/efile   
28            28  http://www.irs.gov/efile   
29            29  http://www.irs.gov/efile   
...          ...                       ...   
6022        6022  http://www.irs.gov/efile   
6023        6023  http://www.irs.gov/efile   
6024        6024  http://www.irs.gov/efile   
6025        6025  http://www.irs.gov/efile   
6026        6026  http://www.irs.gov/efile   
6027        6027  http://www.irs.gov/efile   
6028        6028  http://www.irs.gov/efile   
6029        6029  http://www.irs.gov/efile   
6030        6030  http://www.irs.gov/efile   
6031        6031  http://www.irs.gov/efile   
6032        6032  http://www.irs.gov/efile   
6033        6033  http://www.irs.gov/efile   
6034        6034  http://www.irs.gov/efile   
6035        6035  http://www.irs.gov/efile   
6036        6036  http://www.irs.gov/efile   
6037        6037  http://www.irs.gov/efile   
6038        6038  http://www.irs.gov/efile   
6039        6039  http://www.irs.gov/efile   
6040        6040  http://www.irs.gov/efile   
6041        6041  http://www.irs.gov/efile   
6042        6042  http://www.irs.gov/efile   
6043        6043  http://www.irs.gov/efile   
6044        6044  http://www.irs.gov/efile   
6045        6045  http://www.irs.gov/efile   
6046        6046  http://www.irs.gov/efile   
6047        6047  http://www.irs.gov/efile   
6048        6048  http://www.irs.gov/efile   
6049        6049  http://www.irs.gov/efile   
6050        6050  http://www.irs.gov/efile   
6051        6051  http://www.irs.gov/efile   

                              Return.@xmlns:xsi Return.@returnVersion  \
0     http://www.w3.org/2001/XMLSchema-instance              2017v2.3   
1     http://www.w3.org/2001/XMLSchema-instance              2012v2.3   
2     http://www.w3.org/2001/XMLSchema-instance              2010v3.4   
3                                           NaN              2012v2.1   
4     http://www.w3.org/2001/XMLSchema-instance              2012v2.1   
5     http://www.w3.org/2001/XMLSchema-instance              2012v2.1   
6     http://www.w3.org/2001/XMLSchema-instance              2012v2.1   
7     http://www.w3.org/2001/XMLSchema-instance              2012v2.2   
8     http://www.w3.org/2001/XMLSchema-instance              2012v2.0   
9     http://www.w3.org/2001/XMLSchema-instance              2012v2.1   
10    http://www.w3.org/2001/XMLSchema-instance              2012v2.2   
11                                          NaN              2012v2.1   
12    http://www.w3.org/2001/XMLSchema-instance              2010v3.4   
13    http://www.w3.org/2001/XMLSchema-instance              2012v2.3   
14    http://www.w3.org/2001/XMLSchema-instance           

In [6]:
database_key_list = []

for each in data.keys():
    database_key_list.append(each)

In [8]:
len(database_key_list)

374

In [9]:
database_key_list

['Unnamed: 0',
 'Return.@xmlns',
 'Return.@xmlns:xsi',
 'Return.@returnVersion',
 'Return.ReturnHeader.@binaryAttachmentCnt',
 'Return.ReturnHeader.ReturnTs',
 'Return.ReturnHeader.TaxPeriodEndDt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmEIN',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmName.BusinessNameLine1Txt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.AddressLine1Txt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.CityNm',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.StateAbbreviationCd',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.ZIPCd',
 'Return.ReturnHeader.ReturnTypeCd',
 'Return.ReturnHeader.TaxPeriodBeginDt',
 'Return.ReturnHeader.Filer.EIN',
 'Return.ReturnHeader.Filer.BusinessName.BusinessNameLine1Txt',
 'Return.ReturnHeader.Filer.BusinessName.BusinessNameLine2Txt',
 'Return.ReturnHeader.Filer.BusinessNameControlTxt',
 'Return.ReturnHeader.Filer.PhoneNum',
 'Return.ReturnHeader.Filer.USAddress.AddressLine1Txt',

In [2]:
monroe_html = 'https://s3.amazonaws.com/irs-form-990/201911159349300636_public.xml'
monroe_baseline = xmltodict.parse(requests.get(monroe_html).content)
monroe_dictionary = dict(zip(get_node_names(monroe_baseline),get_node_values(monroe_baseline)))
key_list = list(monroe_dictionary.keys())

In [13]:
key_list

['Return.@xmlns',
 'Return.@xmlns:xsi',
 'Return.@xsi:schemaLocation',
 'Return.@returnVersion',
 'Return.ReturnHeader.@binaryAttachmentCnt',
 'Return.ReturnHeader.ReturnTs',
 'Return.ReturnHeader.TaxPeriodEndDt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmEIN',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmName.BusinessNameLine1Txt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.AddressLine1Txt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.CityNm',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.StateAbbreviationCd',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.ZIPCd',
 'Return.ReturnHeader.ReturnTypeCd',
 'Return.ReturnHeader.TaxPeriodBeginDt',
 'Return.ReturnHeader.Filer.EIN',
 'Return.ReturnHeader.Filer.BusinessName.BusinessNameLine1Txt',
 'Return.ReturnHeader.Filer.BusinessNameControlTxt',
 'Return.ReturnHeader.Filer.PhoneNum',
 'Return.ReturnHeader.Filer.USAddress.AddressLine1Txt',
 'Return.ReturnHeader.Filer.USAddress.CityNm',


# see which keys were not included in data

In [21]:
missing_keys = []
for key in key_list:
    if key not in database_key_list:
        missing_keys.append(key)

In [22]:
len(missing_keys)

179

In [17]:
len(key_list)

470

In [18]:
missing_keys

['Return.@xsi:schemaLocation',
 'Return.ReturnData.IRS990.USAddress.AddressLine1Txt',
 'Return.ReturnData.IRS990.USAddress.CityNm',
 'Return.ReturnData.IRS990.USAddress.StateAbbreviationCd',
 'Return.ReturnData.IRS990.USAddress.ZIPCd',
 'Return.ReturnData.IRS990.TotalVolunteersCnt',
 'Return.ReturnData.IRS990.NetUnrelatedBusTxblIncmAmt',
 'Return.ReturnData.IRS990.PYInvestmentIncomeAmt',
 'Return.ReturnData.IRS990.PYGrantsAndSimilarPaidAmt',
 'Return.ReturnData.IRS990.PYBenefitsPaidToMembersAmt',
 'Return.ReturnData.IRS990.PYTotalProfFndrsngExpnsAmt',
 'Return.ReturnData.IRS990.InfoInScheduleOPartIIIInd',
 'Return.ReturnData.IRS990.GrantAmt',
 'Return.ReturnData.IRS990.ProgSrvcAccomActy2Grp.ExpenseAmt',
 'Return.ReturnData.IRS990.ProgSrvcAccomActy2Grp.RevenueAmt',
 'Return.ReturnData.IRS990.ProgSrvcAccomActy2Grp.Desc',
 'Return.ReturnData.IRS990.GrantsToOrganizationsInd.@referenceDocumentId',
 'Return.ReturnData.IRS990.GrantsToOrganizationsInd.#text',
 'Return.ReturnData.IRS990.GrantsT

# That is weird:

The number of missing keys should be 96, because 470 -374 = 96. But instead it says there are 179 missing keys. Check the number of that are in both lists 

In [24]:
keys_shared = []
for key in key_list:
    if key in database_key_list:
        keys_shared.append(key)

In [25]:
len(keys_shared)

291

# 291 + 179 = 470. So there are even less shared keys than I thought
What are those other keys in the database then?

In [26]:
mystery_keys = []
for key in database_key_list:
    if key not in keys_shared:
        mystery_keys.append(key)

In [27]:
len(mystery_keys)

83

In [28]:
mystery_keys

['Unnamed: 0',
 'Return.ReturnHeader.Filer.BusinessName.BusinessNameLine2Txt',
 'Return.ReturnHeader.PreparerPersonGrp.PreparationDt',
 'Return.ReturnHeader.PreparerPersonGrp.SelfEmployedInd',
 'Return.ReturnData.IRS990.AddressChangeInd',
 'Return.ReturnData.IRS990.GrantsToOrganizationsInd',
 'Return.ReturnData.IRS990.GrantsToIndividualsInd',
 'Return.ReturnData.IRS990.DeductibleNonCashContriInd',
 'Return.ReturnData.IRS990.DeductibleArtContributionInd',
 'Return.ReturnData.IRS990.EmploymentTaxReturnsFiledInd',
 'Return.ReturnData.IRS990.Form8899Filedind',
 'Return.ReturnData.IRS990.Form1098CFiledInd',
 'Return.ReturnData.IRS990.BooksInCareOfDetail.PersonNm',
 'Return.ReturnData.IRS990.ProgramServiceRevenueGrp.ExclusionAmt',
 'Return.ReturnData.IRS990.FundraisingGrossIncomeAmt',
 'Return.ReturnData.IRS990.FundraisingDirectExpensesAmt',
 'Return.ReturnData.IRS990.NetIncmFromFundraisingEvtGrp.TotalRevenueColumnAmt',
 'Return.ReturnData.IRS990.NetIncmFromFundraisingEvtGrp.ExclusionAmt',
 

# so my question is, how did these get in here...
The code that I'm thinking about is:

```Python
for key in key_list:
    new_dict[key] = temp_dict.get(key)
```

the key_list doesn't change. and if the new_dict will always follow flat dictionary pattern. But is it behaving the way I intended? Double check this. Use a couple different affiliate filing addresses

In [29]:
with open('affiliates annual links', 'rb') as infile:
    html_list = pickle.load(infile)

In [37]:
def key_getter(html_link):
    temp_data = xmltodict.parse(requests.get(html_link).content)
    node_list = get_node_names(temp_data)
    return node_list


# ah, just build that comparison function

In [4]:
def shared_keys(keysfeed1, keysfeed2):
    output_keys = []
    for key in keysfeed1:
        if key in keysfeed2:
            output_keys.append(key)
    return output_keys

In [42]:
examine_keys_1 = key_getter(html_list[2])
testcase = shared_keys(examine_keys_1, key_list)

In [44]:
testcase

['Return.@xmlns',
 'Return.@returnVersion',
 'Return.ReturnHeader.Filer.EIN',
 'Return.ReturnHeader.BuildTS',
 'Return.ReturnData.IRS990.@documentId',
 'Return.ReturnData.IRS990ScheduleA.@documentId',
 'Return.ReturnData.IRS990ScheduleD.@documentId',
 'Return.ReturnData.IRS990ScheduleO.@documentId']

# this link is from 2013 filing and only shares 8 keys. Not acceptable

# what about from later years (just pick a high number)

In [45]:
testcase2 = shared_keys(key_getter(html_list[3000]), key_list)

In [47]:
len(testcase2)

305

# better, but this still doesn't explain where those other  keys come from. Check what happens from the output of the dict(zip()) stuff

In [48]:
temp_data = xmltodict.parse(requests.get(html_list[2]).content)
node_list = get_node_names(temp_data)
node_values = get_node_values(temp_data)
temp_dict = dict(zip(node_list,node_values))

In [49]:
temp_dict

{'Return.@xmlns': 'http://www.irs.gov/efile',
 'Return.@returnVersion': '2012v2.1',
 'Return.ReturnHeader.@binaryAttachmentCount': '0',
 'Return.ReturnHeader.Timestamp': '2013-09-30T07:24:40-04:00',
 'Return.ReturnHeader.TaxPeriodEndDate': '2013-06-30',
 'Return.ReturnHeader.PreparerFirm.EIN': '570826140',
 'Return.ReturnHeader.PreparerFirm.PreparerFirmBusinessName.BusinessNameLine1': 'Gibbons & Usry CPAs PA',
 'Return.ReturnHeader.PreparerFirm.PreparerFirmUSAddress.AddressLine1': 'PO Box 489 113 S Brooks St',
 'Return.ReturnHeader.PreparerFirm.PreparerFirmUSAddress.City': 'Manning',
 'Return.ReturnHeader.PreparerFirm.PreparerFirmUSAddress.State': 'SC',
 'Return.ReturnHeader.PreparerFirm.PreparerFirmUSAddress.ZIPCode': '29102',
 'Return.ReturnHeader.ReturnType': '990',
 'Return.ReturnHeader.TaxPeriodBeginDate': '2012-07-01',
 'Return.ReturnHeader.Filer.EIN': '571015546',
 'Return.ReturnHeader.Filer.Name.BusinessNameLine1': 'HABITAT FOR HUMANITY INTL CLARENDON',
 'Return.ReturnHeader.Fi

In [53]:
len(temp_dict.keys())

314

# Time to test theory:
The following...

```Python
for key in key_list:
    new_dict[key] = temp_dict.get(key)
```
tested on early affiliate links should produce a new dictionary with the same keys as the monroe county dictionary, but with empty values

In [50]:
new_dict = {}
for key in key_list:
    new_dict[key] = temp_dict.get(key)

In [51]:
new_dict

{'Return.@xmlns': 'http://www.irs.gov/efile',
 'Return.@xmlns:xsi': None,
 'Return.@xsi:schemaLocation': None,
 'Return.@returnVersion': '2012v2.1',
 'Return.ReturnHeader.@binaryAttachmentCnt': None,
 'Return.ReturnHeader.ReturnTs': None,
 'Return.ReturnHeader.TaxPeriodEndDt': None,
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmEIN': None,
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmName.BusinessNameLine1Txt': None,
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.AddressLine1Txt': None,
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.CityNm': None,
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.StateAbbreviationCd': None,
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.ZIPCd': None,
 'Return.ReturnHeader.ReturnTypeCd': None,
 'Return.ReturnHeader.TaxPeriodBeginDt': None,
 'Return.ReturnHeader.Filer.EIN': '571015546',
 'Return.ReturnHeader.Filer.BusinessName.BusinessNameLine1Txt': None,
 'Return.ReturnHeader.Filer.BusinessNameControlTxt': None,

In [52]:
len(new_dict.keys())

470

# issue isn't here, at least not with the early affiliate links. What about with affiliates that have a different filing type? Again, it shouldn't matter... right?

# Wait... I figured it out. I accidently used the node_list of the last affiliate, rather than monroes.

# Still need to figure out if 2014 is in line with the rest of the reports

In [3]:
monroe_html = 'https://s3.amazonaws.com/irs-form-990/201911159349300636_public.xml'
monroe_baseline = xmltodict.parse(requests.get(monroe_html).content)
monroe_dictionary = dict(zip(get_node_names(monroe_baseline),get_node_values(monroe_baseline)))
key_list = list(monroe_dictionary.keys())

In [5]:
test_html = 'https://s3.amazonaws.com/irs-form-990/201411709349300241_public.xml'
test_case = xmltodict.parse(requests.get(test_html).content)
test_keys = get_node_names(test_case)

In [6]:
test_keys

['Return.@xmlns',
 'Return.@xmlns:xsi',
 'Return.@xsi:schemaLocation',
 'Return.@returnVersion',
 'Return.ReturnHeader.@binaryAttachmentCnt',
 'Return.ReturnHeader.ReturnTs',
 'Return.ReturnHeader.TaxPeriodEndDt',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmEIN',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerFirmName.BusinessNameLine1',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.AddressLine1',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.City',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.State',
 'Return.ReturnHeader.PreparerFirmGrp.PreparerUSAddress.ZIPCode',
 'Return.ReturnHeader.ReturnTypeCd',
 'Return.ReturnHeader.TaxPeriodBeginDt',
 'Return.ReturnHeader.Filer.EIN',
 'Return.ReturnHeader.Filer.BusinessName.BusinessNameLine1',
 'Return.ReturnHeader.Filer.BusinessNameControlTxt',
 'Return.ReturnHeader.Filer.PhoneNum',
 'Return.ReturnHeader.Filer.USAddress.AddressLine1',
 'Return.ReturnHeader.Filer.USAddress.City',
 'Return.ReturnHeader.Filer.

In [8]:
similiar_keys = shared_keys(key_list, test_keys)

In [10]:
len(similiar_keys)

253

# That signifies that the xml keys have the same form from 2014 onward.
# One last thing: what about those other filing types? start with 990O (the weird one)

In [3]:
#201510449349300426
test_html1 = 'https://s3.amazonaws.com/irs-form-990/201510449349300426_public.xml'
test_case1 = xmltodict.parse(requests.get(test_html1).content)
test_keys1 = get_node_names(test_case1)


In [8]:
k = shared_keys(test_keys1, key_list)

In [9]:
k

['Return.@xmlns',
 'Return.@xsi:schemaLocation',
 'Return.@xmlns:xsi',
 'Return.@returnVersion',
 'Return.ReturnHeader.@binaryAttachmentCnt',
 'Return.ReturnHeader.ReturnTs',
 'Return.ReturnHeader.TaxPeriodEndDt',
 'Return.ReturnHeader.ReturnTypeCd',
 'Return.ReturnHeader.TaxPeriodBeginDt',
 'Return.ReturnHeader.Filer.EIN',
 'Return.ReturnHeader.Filer.BusinessNameControlTxt',
 'Return.ReturnHeader.Filer.PhoneNum',
 'Return.ReturnHeader.BusinessOfficerGrp.PersonNm',
 'Return.ReturnHeader.BusinessOfficerGrp.PersonTitleTxt',
 'Return.ReturnHeader.BusinessOfficerGrp.PhoneNum',
 'Return.ReturnHeader.BusinessOfficerGrp.SignatureDt',
 'Return.ReturnHeader.BusinessOfficerGrp.DiscussWithPaidPreparerInd',
 'Return.ReturnHeader.PreparerPersonGrp.PreparerPersonNm',
 'Return.ReturnHeader.PreparerPersonGrp.PTIN',
 'Return.ReturnHeader.PreparerPersonGrp.PhoneNum',
 'Return.ReturnHeader.TaxYr',
 'Return.ReturnHeader.BuildTS',
 'Return.ReturnData.IRS990.@documentId',
 'Return.ReturnData.IRS990.@referen

In [10]:
len(k)

259

# that shouldn't be a problem. What about 990EZ?

In [11]:
test_html2 = 'https://s3.amazonaws.com/irs-form-990/201520129349200512_public.xml'
test_case2 = xmltodict.parse(requests.get(test_html2).content)
test_keys2 = get_node_names(test_case2)

In [12]:
k2 = shared_keys(test_keys2, key_list)

In [14]:
len(k2)

46

# absolutely drop 990EZ. Are there any other filing types? 
# Yes there is 990EO

In [15]:
# 201901079349200630
test_html3 = 'https://s3.amazonaws.com/irs-form-990/201901079349200630_public.xml'
test_case3 = xmltodict.parse(requests.get(test_html3).content)
test_keys3 = get_node_names(test_case3)

In [16]:
k3 = shared_keys(test_keys3, key_list)

In [18]:
len(k3)

41

# Drop 990EO, its not worth it